cmd/transcriber/call/live_captions.go

```diff
@@ -0,0 +1,462 @@
+		window, prevTranscribedPos = cutWindowToSize(ctx.trackID, window, segments, prevTranscribedPos)
+		prevWindowLen = len(window)
+
+		// Use a for loop and a select so that we can drop ticks waiting for the transcriber.
+	waitForTranscription:
+		for {
+			select {
+			case <-ticker.C:
+				slog.Debug("processLiveCaptionsForTrack: dropped a tick waiting for the transcriber",
+					slog.String("trackID", ctx.trackID))
+			case text := <-transcribedCh:
+				if len(text) == 0 {
+					// Either transcribedCh was closed above (captionQueueCh full), or audio transcription failed.
+					// Note: this appears to happen when the transcriber fails to decode a block of audio.
+					// Usually the probability returned for the language is very low, which makes sense.
+					slog.Debug("processLiveCaptionsForTrack: received empty text, ignoring.")
+					break waitForTranscription
+				}
+				if err := t.client.SendWs(wsEvCaption, public.CaptionMsg{
+					SessionID:     ctx.sessionID,
+					UserID:        ctx.user.Id,
+					Text:          text,
+					NewAudioLenMs: float64(newAudioLenMs),
+				}, false); err != nil {
+					slog.Error("processLiveCaptionsForTrack: error sending ws captions",
+						slog.String("err", err.Error()),
+						slog.String("trackID", ctx.trackID))
+				}
+
+				break waitForTranscription
+			}
+		}
+	}
+}
+
+type segmentSamples struct {
+	Start   int
+	End     int
+	Silence bool
+}
+
+// convertToSegmentSamples turns the speech.Segments (in time) into segmentSamples (measured in samples)
+func convertToSegmentSamples(segments []speech.Segment, audioLen int) []segmentSamples {
+	// close off the last segment (we know segments > 0)
+	if segments[len(segments)-1].SpeechEndAt == 0 {
+		segments[len(segments)-1].SpeechEndAt = float64(audioLen) / trackOutAudioRate
+	}
+
+	var ret []segmentSamples
+	lastEndAtIdx := 0
+	for _, seg := range segments {
+		start := int(seg.SpeechStartAt * trackOutAudioRate)
+		end := int(seg.SpeechEndAt * trackOutAudioRate)
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     start,
+			Silence: true,
+		})
+		ret = append(ret, segmentSamples{
+			Start:   start,
+			End:     end,
+			Silence: false,
+		})
+		lastEndAtIdx = end
+	}
+
+	if lastEndAtIdx < audioLen {
+		ret = append(ret, segmentSamples{
+			Start:   lastEndAtIdx,
+			End:     audioLen,
+			Silence: true,
+		})
+	}
+
+	return ret
+}
+
+// removeShortSpeeches removes small sections of speech because either they are not actual words,
+// or the transcriber will have trouble with such a short amount.
+func removeShortSpeeches(segments []segmentSamples) []segmentSamples {
+	for i, seg := range segments {
+		if !seg.Silence && (seg.End-seg.Start) < minSpeechLengthSamples {
+			segments[i].Silence = true
+		}
+	}
+	return segments
+}
+
+func cleanAudio(audio []float32, segments []segmentSamples) []float32 {
+	cleaned := append([]float32(nil), audio...)
```

Comment:
Wondering if it'd be possible to do this in place instead of allocating every time.

Reply:
I had it like that, but then went this way because I didn't want to permanently change the window for subsequent ticks. I'm not worried about the small memory cost, when you compare it to the cost of the transcriber.